hal/network/ncp_client/sara/sara_ncp_client.cpp

```diff
@@ -1533,12 +1542,24 @@ int SaraNcpClient::configureApn(const CellularNetworkConfig& conf) {
 
         // If failed above i.e., netConf_ is still not valid, look for network settings based on IMSI
         if (!netConf_.isValid()) {
-            memset(buf, 0, sizeof(buf));
-            auto resp = parser_.sendCommand(UBLOX_CIMI_TIMEOUT, "AT+CIMI");
-            CHECK_PARSER(resp.readLine(buf, sizeof(buf)));
-            const int r = CHECK_PARSER(resp.readResult());
-            CHECK_TRUE(r == AtResponse::OK, SYSTEM_ERROR_AT_NOT_OK);
-            netConf_ = networkConfigForImsi(buf, strlen(buf));
+            int imsiCount = 0;
+            do {
+                auto resp = parser_.sendCommand(UBLOX_CIMI_TIMEOUT, "AT+CIMI");
+                memset(buf, 0, sizeof(buf));
+                int imsiLength = 0;
+                if (resp.hasNextLine()) {
+                    imsiLength = CHECK(resp.readLine(buf, sizeof(buf)));
+                }
+                const int r = CHECK_PARSER(resp.readResult());
+                if (r == AtResponse::OK && imsiLength > 0) {
+                    netConf_ = networkConfigForImsi(buf, strlen(buf));
+                    break;
+                } else if (imsiCount >= IMSI_MAX_RETRY_CNT-1) {
+                    // if max retries are exhausted
+                    return SYSTEM_ERROR_AT_RESPONSE_UNEXPECTED;
+                }
+                HAL_Delay_Milliseconds(250*imsiCount);
+            } while (++imsiCount < IMSI_MAX_RETRY_CNT);
```

Comment:
We have pretty much two identical pieces of code that implement this retry logic. It sounds like a good idea to decouple IMSI querying into a separate method and keep this retry logic in a single place.